### ISSUE TITLE
List to CSV blocks no longer produce values that can be inserted into Fusion Tables

### ISSUE THREAD

**Issue by bennytheshap** [mit-cml/appinventor-sources#848]
# tl;dr

FusionTablesControl.InsertRow expects its values to be single quoted, but list-to-csv-row produces double quoted values. 
# suggested fix

FusionTablesControl should accept a variety of parameter types (e.g. lists, double quoted strings, etc.) coercing its parameters into whatever form it needs.
# more detail

**Scenario**: Stacy wishes to record the names and favorite foods of her friends into a Fusion Table. She creates an app that has TextBoxes called NameBox and FoodBox.
Her friend @marksherman enters the strings  _Mark_ and _beer_ in the NameBox and FoodBox, respectively, and presses a button.

**What Happens**: The following code does not work:
<img width="784" alt="screen shot 2016-09-27 at 5 53 25 pm" src="https://cloud.githubusercontent.com/assets/122056/18896295/ce7e0c7a-84dd-11e6-93aa-f0b723104b9f.png">

App Inventor throws the following error at run time:
![2016-09-27 17 41 56](https://cloud.githubusercontent.com/assets/122056/18896333/2228d120-84de-11e6-97ac-c7b66b18ffd5.jpg)

**What Does Work But Shouldn't Be Necessary**:
<img width="1051" alt="screen shot 2016-09-27 at 6 00 27 pm" src="https://cloud.githubusercontent.com/assets/122056/18896398/a792e788-84de-11e6-9cc7-9c45113b820c.png">

**Why This Behavior Is Problematic**
It would be easier for students if the various App Inventor library functions were more consistent in the types and formats of the parameters that they expect and the values that they produce. This would make it easier for students to combine library functions into more sophisticated programs without extra cognitive overhead.

**Comment by bennytheshap** [comment]
Tagging @miketissenbaum in case he wants me to send a simple .aia.

**Comment by ewpatton** [comment]
Looking at the git logs for CsvUtil, which performs the translation from YailList into CSV, the code has not been touched since 2013. If this ever worked, it was by the grace of Google that FusionTables accepted double quotes, and at some point the behavior was then changed to accept only single quotes. Therefore this is not a regression vis-a-vis the list to csv primitive.

That being said, I do agree that it would be better if the FusionTablesControl was more robust to these issues and the number of questions asked on the forums support that sentiment. The general issue with FusionTablesControl is that while App Inventor does a relatively good job of hiding the complexities of Android, Java, etc., the FusionTablesControl basically requires the user to gain some amount of understanding of the syntax and semantics of SQL.

In terms of a resolution, at the extreme end of the spectrum I envision having an SQL blocks section that can be used to eliminate all of the SQL syntax errors by bringing it into Blockly rather than having users perform the string concatenation dance. A happy middle ground might be allowing commands like INSERT to accept a list of values rather than text ([per AI2 documentation](http://ai2.appinventor.mit.edu/reference/components/storage.html#FusionTablesControl)), which conceptually the VALUES part of the INSERT is anyway. There is also the cost of the semantic change to the block and its affect on existing projects using the FusionTablesControl.

I've marked this as needs discussion and welcome any comments.

**Comment by josmas** [comment]
I also agree this is an issue, but wonder how many people would create a _csv list_ instead of _joining_ values directly as in the screenshot:

<img width="553" alt="ft_insertrow" src="https://cloud.githubusercontent.com/assets/194925/18908829/17dbc76e-8569-11e6-9cc6-96077884b1c4.png">

If you look at Taifun's examples, he uses procedures like `quotify = join + replace all text`, and has a good collection of tips on [how to look for Errors in FT](http://puravidaapps.com/fusionerror.php), which clearly shows how raw and low level the component is.

A compromise could be a new block _InsertRowFromList_ that takes _lists_ for both column names and values. It would be an addition to the component, so no need to fuss about with existing projects, but having several blocks for the same thing could be confusing.

Are there any examples of blocks that can take several types? On the Blockly side it's easy to do multiple type checks, but this is generated from Java... could the block take an _Object_ parameter and then coerce the type?

**Comment by bennytheshap** [comment]
@josmas Yes, the manual string concatenation approach is prevalent in examples, but is problematic. For example, the screenshot above would break if the app's user includes an apostrophe in their form input. Some of Taifun's examples include string escaping, which would prevent that particular error. But we should consider: what are we trying to accomplish, and are these abstractions the right ones? Something that accepts lists, and then does whatever escaping is necessary might make it easier for learners to reason about their programs. 

Consider an analogy: the argument for blocks instead of character-based text editing is that we want learners to be able to focus on the semantics of their programs without getting tripped up in syntactical complexity. Generally, App Inventor does this really well. The string construction and escaping approach in the above screenshot and in Taifun's examples runs briskly in the opposite direction!

**Comment by halatmit** [comment]
Thanks for filing this as an issue


== Hal

Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.

**Comment by josmas** [comment]
@bennytheshap I totally agree on that, the abstractions are not appropriate, and as mentioned in the previous message, Taifun is merely patching up blocks that don't behave on a par with other blocks.

**Comment by ewpatton** [comment]
Closing because FusiontablesControl is no longer supported.
